src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java

```diff
@@ -270,27 +351,51 @@ protected SCMRevision retrieve(SCMHead head, TaskListener listener) throws IOExc
-        listener.error("Please implement " + getClass().getName() + ".doRetrieve(SCMHead, TaskListener, GHRepository)");
-        SCMHeadObserver.Selector selector = SCMHeadObserver.select(head);
-        doRetrieve(selector, listener, repo);
-        return selector.result();
+    protected SCMRevision doRetrieve(SCMHead head, TaskListener listener, GHRepository repo) throws IOException, InterruptedException {
+        GHRef ref;
+        if (head instanceof PullRequestSCMHead) {
+            int number = ((PullRequestSCMHead) head).getNumber();
+            ref = repo.getRef("pull/" + number + "/merge");
+        } else {
+            ref = repo.getRef("heads/" + head.getName());
+        }
+        return new SCMRevisionImpl(head, ref.getObject().getSha());
     }
 
-    public static abstract class AbstractGitHubSCMSourceDescriptor extends SCMSourceDescriptor {
+    @Extension public static class DescriptorImpl extends SCMSourceDescriptor {
 
-        private static final Logger LOGGER = Logger.getLogger(AbstractGitHubSCMSourceDescriptor.class.getName());
+        private static final Logger LOGGER = Logger.getLogger(DescriptorImpl.class.getName());
 
         public static final String defaultIncludes = "*";
         public static final String defaultExcludes = "";
         public static final String ANONYMOUS = "ANONYMOUS";
         public static final String SAME = "SAME";
 
-        public FormValidation doCheckScanCredentialsId(@QueryParameter String value) {
-            if (!value.isEmpty()) {
-                return FormValidation.ok();
+        @Override
+        public String getDisplayName() {
+            return "GitHub";
+        }
+
+        @Restricted(NoExternalUse.class)
+        public FormValidation doCheckScanCredentialsId(@AncestorInPath SCMSourceOwner context,
+                @QueryParameter String scanCredentialsId, @QueryParameter String apiUri) {
+            if (!scanCredentialsId.isEmpty()) {
+                StandardCredentials credentials = Connector.lookupScanCredentials(context, apiUri, scanCredentialsId);
+                if (credentials == null) {
+                    FormValidation.error("Invalid credentials");
+                } else {
+                    try {
+                        GitHub connector = Connector.connect(apiUri, credentials);
+                        if (connector.isCredentialValid()) {
+                            return FormValidation.ok();
+                        }
+                    } catch (IOException e) {
+                        // ignore, never thrown
+                    }
+                }
+                return FormValidation.error("Invalid credentials");
             } else {
-                return FormValidation.warning("Credentials are required");
+                return FormValidation.warning("Credentials are recommended");
```

Comment:
I prefer to show a message if the notifications did not happen by not use credentials.